lib/elixir/lib/enum.ex

```diff
@@ -2358,13 +2358,17 @@ defmodule Enum do
 
   ## Examples
 
-      iex> Enum.with_index [:a, :b, :c]
+      iex> Enum.with_index([:a, :b, :c])
       [a: 0, b: 1, c: 2]
 
+      iex> Enum.with_index([:a, :b, :c], 3)
+      [a: 3, b: 4, c: 5]
+
   """
-  @spec with_index(t) :: [{element, non_neg_integer}]
-  def with_index(enumerable) do
-    map_reduce(enumerable, 0, fn x, acc ->
+  @spec with_index(t) :: [{element, integer}]
+  def with_index(enumerable), do: with_index(enumerable, 0)
+  def with_index(enumerable, offset) do
```

Comment:
The benefit is what it would be considered one function, which is useful in the documentation. For example, in your current patch, you are only documenting `with_index/1`, `with_index/2` would show with no documentation.